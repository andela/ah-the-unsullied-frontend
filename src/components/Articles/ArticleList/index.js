import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';
import _ from 'lodash';
import Pagination from 'react-js-pagination';
import { Col, Preloader, Row } from 'react-materialize';

import { getArticles } from '../../../actions/ArticleActions/actions';
import '../../../assets/css/articles.scss';

class ArticleList extends Component {
  constructor() {
    super();
    this.state = {
      activePage: 1,
      loaded: false
    };
  }
  componentWillMount() {
    let arr = _.values(this.props.articles['results']);
   if(arr.length === 0){
   this.props.getArticles();
   }
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.articles['count'] > 0) {
      this.setState({
        loaded: true
      });
      window.scrollTo(0, 0);
    }
  }
  handlePageChange = pageNumber => {
    this.setState({
      loaded: false
    });
    this.props.getArticles(pageNumber);
    this.setState({ activePage: pageNumber });
  };
  articleCard() {
    const arr = _.values(this.props.articles['results']);
    const articles = arr.map(article => {
      const date = article.created_at.slice(0, 10);
      const slug = article.slug;

      return (
        <div className="row">
          <div className="col s12 m8">
            <div className="card darken-1">
              <div className="card-content black-text">
                <span id="something" className="card-title teal-text">
                  <Link
                    to={{
                      pathname: `/article/${slug}`,
                      slug: { slug },
                      single: { ...article }
                    }}
                  >
                    {article.title}
                  </Link>
                </span>
                <p className="paragraph-text grey-text darken-2">
                  {article.description}
                </p>
              </div>
              <p className="author-text grey-text">
                by <span className="teal-text">{article.author.username}</span>
              </p>
              <div className="dateinfo">
                <i className="material-icons">date_range</i>
                <span className="grey-text">
                  {date} | {article.read_time}
                </span>
                <div className="card-action">
                  <i className="fa fa-thumbs-up" aria-hidden="true">
                    {article.likes}
                  </i>
                  &nbsp;&nbsp;&nbsp;
                  <i className="fa fa-thumbs-down" aria-hidden="true">
                    {article.dislikes}
                  </i>
                </div>
              </div>
            </div>
          </div>
        </div>
      );
    });
    return (
      <div>
        {articles}
        {this.props.articles['count'] > 10 && (
          <div id="pagination">
            <Pagination
              prevPageText="prev"
              nextPageText="next"
              firstPageText="first"
              lastPageText="last"
              activePage={this.state.activePage}
              itemsCountPerPage={10}
              totalItemsCount={this.props.articles['count']}
              pageRangeDisplayed={Math.ceil(this.props.articles['count'] / 10)}
              onChange={this.handlePageChange}
            />
          </div>
        )}
      </div>
    );
  }
  render() {
    if (!this.state.loaded) {
      return (
        <div className="container">
          <Row>
            <Col s={5}>
              <Preloader size="big" />
            </Col>
          </Row>
        </div>
      );
    } else {
      return <div className="">{this.articleCard()}</div>;
    }
  }
}

ArticleList.propTypes = {
  getArticles: PropTypes.func.isRequired,
  articles: PropTypes.array.isRequired
};
ArticleList.defaultProps = {
  articles: PropTypes.array.isRequired
};
const mapStateToProps = state => {
  return state.articles;
};
export default withRouter(
  connect(
    mapStateToProps,
    { getArticles }
  )(ArticleList)
);
