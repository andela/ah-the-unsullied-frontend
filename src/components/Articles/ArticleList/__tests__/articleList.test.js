import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16/build';
import ArticleList from '..';

Enzyme.configure({ adapter: new Adapter() });

describe('ArticleList', () => {
  it('should render correctly', () => {
    const component = shallow(<ArticleList />);
    expect(component).toMatchSnapshot();
  });
});
