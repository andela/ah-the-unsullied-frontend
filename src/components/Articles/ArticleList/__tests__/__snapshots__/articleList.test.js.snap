// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article list view should render article list 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleList
    article={
      Object {
        "author": Object {
          "bio": "i have abio",
          "image": "hshhshs",
          "username": "nesh",
        },
        "body": "gsgsgsg",
        "created_at": "01-01-2019",
        "description": "some stuff",
        "dislikes": 1,
        "displiked": false,
        "is_reported": false,
        "liked": false,
        "likes": 2,
        "slug": "my-article",
        "title": "my article",
      }
    }
    articleCard={[MockFunction]}
    articles={
      Object {
        "articles": Object {
          "count": 12,
          "pages": Object {
            "next_page": Object {},
            "previous_page": Object {},
          },
          "resuts": Array [
            Object {
              "author": Object {
                "bio": "i have abio",
                "image": "hshhshs",
                "username": "nesh",
              },
              "body": "gsgsgsg",
              "created_at": "01-01-2019",
              "description": "some stuff",
              "dislikes": 1,
              "displiked": false,
              "is_reported": false,
              "liked": false,
              "likes": 2,
              "slug": "my-article",
              "title": "my article",
            },
            Object {
              "author": Object {
                "bio": "i have abio",
                "image": "hshhshs",
                "username": "nesh",
              },
              "body": "gsgsgsg",
              "created_at": "01-01-2019",
              "description": "some stuff",
              "dislikes": 1,
              "displiked": false,
              "is_reported": false,
              "liked": false,
              "likes": 2,
              "slug": "my-article",
              "title": "my article",
            },
          ],
        },
      }
    }
    auth={
      Object {
        "isAuthenticated": true,
        "user": Object {
          "username": "nesh",
        },
      }
    }
    getArticles={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    match={
      Object {
        "params": Object {
          "slug": "my article",
        },
      }
    }
    pageNumber={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Row>
        <Col
          s={5}
        >
          <Preloader
            active={true}
            color="blue"
            flashing={false}
            size="big"
          />
        </Col>
      </Row>,
      "className": "container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Col
          s={5}
        >
          <Preloader
            active={true}
            color="blue"
            flashing={false}
            size="big"
          />
        </Col>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Preloader
            active={true}
            color="blue"
            flashing={false}
            size="big"
          />,
          "s": 5,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": true,
            "color": "blue",
            "flashing": false,
            "size": "big",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Row>
          <Col
            s={5}
          >
            <Preloader
              active={true}
              color="blue"
              flashing={false}
              size="big"
            />
          </Col>
        </Row>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Col
            s={5}
          >
            <Preloader
              active={true}
              color="blue"
              flashing={false}
              size="big"
            />
          </Col>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Preloader
              active={true}
              color="blue"
              flashing={false}
              size="big"
            />,
            "s": 5,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": true,
              "color": "blue",
              "flashing": false,
              "size": "big",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
